features/dd-sdk-android-trace/src/main/kotlin/com/datadog/android/trace/internal/RumContextPropagator.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+package com.datadog.android.trace.internal
+
+import com.datadog.android.api.InternalLogger
+import com.datadog.android.api.context.DatadogContext
+import com.datadog.android.api.feature.Feature
+import com.datadog.android.api.feature.FeatureSdkCore
+import com.datadog.android.api.feature.getContextFuture
+import com.datadog.android.lint.InternalApi
+import com.datadog.android.log.LogAttributes
+import com.datadog.android.trace.api.span.DatadogSpan
+import com.datadog.android.trace.api.span.DatadogSpanBuilder
+import com.datadog.trace.core.DDSpan
+import java.util.concurrent.Future
+import java.util.concurrent.TimeUnit
+
+/**
+ * For internal usage only.
+ *
+ * This is a utility class that manages the propagation of RUM contexts for span events and HTTP tracing.
+ */
+@InternalApi
+class RumContextPropagator(private val sdkCoreProvider: () -> FeatureSdkCore?) {
+
+    private val sdkCore: FeatureSdkCore?
+        get() = sdkCoreProvider()
+
+    private val internalLogger: InternalLogger?
+        get() = sdkCore?.internalLogger
+
+    private fun injectRumContext(builder: DatadogSpanBuilder) {
+        sdkCore?.getFeature(Feature.RUM_FEATURE_NAME)
+            ?.getContextFuture(withFeatureContexts = setOf(Feature.RUM_FEATURE_NAME))
+            ?.let { lazyContext -> builder.withTag(DATADOG_INITIAL_CONTEXT, lazyContext) }
+    }
+
+    private fun extractRumContextInternal(instance: Any, block: Boolean = false) {
+        val future = instance.getTag<Future<DatadogContext?>>(DATADOG_INITIAL_CONTEXT)
+
+        if (future != null) {
+            val datadogContext = when {
+                block -> future.getOrNull(CONTEXT_RESOLUTION_TIMEOUT, TimeUnit.SECONDS)
```

Comment:
minor: we may be interested in logging to the telemetry cases when it wasn't resolved on timeout. It would be some signal to investigate (although I'm not sure if it would be easily actionable). WDYT?